packages/lit-dev-server/src/middleware/fake-github-middleware.ts

```diff
@@ -0,0 +1,133 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import type Koa from 'koa';
+
+/**
+ * Options for fakeGitHubMiddleware.
+ */
+export interface FakeGitHubMiddlewareOptions {
+  /**
+   * GitHub OAuth App client ID.
+   *
+   * In reality, this is generated when a GitHub OAuth App is first created. For
+   * this fake, it can be any string.
+   */
+  clientId: string;
+
+  /**
+   * GitHub OAuth app secret.
+   *
+   * In reality, this is generated from the GitHub OAuth App settings screen.
+   * For this fake, it can be any string.
+   */
+  clientSecret: string;
+
+  /**
+   * URL to redirect to after app authentication or installation.
+   *
+   * In reality, this is configured in the GitHub OAuth App settings screen. For
+   * this fake, it can be any URL.
+   */
+  redirectUrl: string;
+}
+
+/**
+ * Koa middleware that simulates various GitHub APIs and authentication screens
+ * used by the lit.dev Playground.
+ */
+export const fakeGitHubMiddleware = (
+  options: FakeGitHubMiddlewareOptions
+): Koa.Middleware => {
+  let fake = new FakeGitHub(options);
+  return async (ctx, next) => {
+    if (ctx.path === '/reset') {
+      return fake.reset(ctx);
+    } else if (ctx.path === '/login/oauth/authorize') {
+      return fake.authorize(ctx);
+    } else {
+      return next();
+    }
+  };
+};
+
+const randomString = () => String(Math.floor(Math.random() * 1e10));
+
+class FakeGitHub {
+  private readonly _options: FakeGitHubMiddlewareOptions;
+  private readonly _userToCode = new Map<string, string>();
+
+  constructor(options: FakeGitHubMiddlewareOptions) {
+    this._options = options;
+  }
+
+  /**
+   * Generate a random user ID for this session and persist it in a cookie.
+   */
+  private _getOrSetUserId(ctx: Koa.Context) {
+    let userId = ctx.cookies.get('userid');
+    if (!userId) {
+      userId = randomString();
+      ctx.cookies.set('userid', userId);
+    }
+    return userId;
+  }
+
+  /**
+   * Clear all server state and browser cookies.
+   */
+  reset(ctx: Koa.Context) {
+    this._userToCode.clear();
+    ctx.cookies.set('userid', null);
+    ctx.cookies.set('authorized', null);
+    ctx.status = 200;
+    ctx.body = 'fake github successfully reset';
+  }
+
+  /**
+   * Simulates https://github.com/login/oauth/authorize, the screen that prompts
+   * a user to authorize a GitHub OAuth App.
+   *
+   * Documentation:
+   * https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#1-request-a-users-github-identity
+   */
+  authorize(ctx: Koa.Context) {
+    if (ctx.query.client_id !== this._options.clientId) {
+      ctx.status = 400;
+      ctx.body = 'error: missing or incorrect client_id url parameter';
+      return;
+    }
+
+    const userId = this._getOrSetUserId(ctx);
+    const code = randomString();
+    this._userToCode.set(userId, code);
+
+    const authorizeUrl = `${this._options.redirectUrl}?code=${code}`;
```

Comment:
Do we want to also support `state` for additional security?

Reply:
Good question. I've been trying to think of whether we can use this for meaningful extra security, and I'm not sure yet.

Let's say our server generated a random string to be put in the `state` parameter, and then checked that `state` parameter when the response eventually comes back. This would seem to guard against attackers sending victims to an authorization URL outside of our login flow. But what would prevent an attacker from calling our API that generates random `states` themselves, and just including that?

One related key thing to know is that it shouldn't be possible to intercept a code by changing the `redirect_uri` parameter to something we don't control, because GitHub always checks that parameter against a URL hard coded into our App settings (it has to match it, or be a subdirectory of it: https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#redirect-urls).